cpp/demo/poisson-matrix-free/main.cpp

```diff
@@ -0,0 +1,209 @@
+
+    // Create mesh and function space
+    auto mesh = std::make_shared<mesh::Mesh>(mesh::create_rectangle(
+        comm, {{{0.0, 0.0}, {1.0, 1.0}}}, {10, 10}, mesh::CellType::triangle,
+        mesh::GhostMode::none));
+
+    auto V = std::make_shared<fem::FunctionSpace>(
+        fem::create_functionspace(functionspace_form_poisson_M, "ui", mesh));
+
+    // Prepare and set Constants for the bilinear form
+    auto f = std::make_shared<fem::Constant<T>>(-6.0);
+
+    // Define variational forms
+    auto L = std::make_shared<fem::Form<T>>(
+        fem::create_form<T>(*form_poisson_L, {V}, {}, {{"f", f}}, {}));
+
+    // Action of the bilinear form "a" on a function ui
+    auto ui = std::make_shared<fem::Function<T>>(V);
+    auto M = std::make_shared<fem::Form<T>>(
+        fem::create_form<T>(*form_poisson_M, {V}, {{"ui", ui}}, {{}}, {}));
+
+    // Define boundary condition
+    auto u_D = std::make_shared<fem::Function<T>>(V);
+    u_D->interpolate(
+        [](auto&& x) {
+          return 1 + xt::square(xt::row(x, 0)) + 2 * xt::square(xt::row(x, 1));
+        });
+
+    auto facets = mesh::exterior_facet_indices(*mesh);
+    const auto bdofs = fem::locate_dofs_topological({*V}, 1, facets);
+    auto bc = std::make_shared<const fem::DirichletBC<T>>(u_D, bdofs);
+
+    // Assemble RHS vector
+    la::Vector<T> b(V->dofmap()->index_map, V->dofmap()->index_map_bs());
+    fem::assemble_vector(b.mutable_array(), *L);
+
+    // Apply lifting to account for Dirichlet boundary condition
+    // b <- b - A * x_bc
+    fem::set_bc(ui->x()->mutable_array(), {bc}, -1.0);
+    dolfinx::fem::assemble_vector(b.mutable_array(), *M);
+
+    // Communicate ghost values
+    b.scatter_rev(common::IndexMap::Mode::add);
+
+    // Set BC dofs to zero (effectively zeroes columns of A)
+    fem::set_bc(b.mutable_array(), {bc}, 0.0);
+
+    // Create function for computing the action of A on x (y = Ax)
+    std::function<void(la::Vector<T>&, la::Vector<T>&)> action
+        = [&](la::Vector<T>& x, la::Vector<T>& y)
+    {
+      // Update ghost values and zero y
+      x.scatter_fwd();
+      y.set(0);
+
+      // Update coefficient ui (just copy data from x to ui)
+      std::copy(x.array().begin(), x.array().end(),
+                ui->x()->mutable_array().begin());
+
+      // Compute action of A on x
+      dolfinx::fem::assemble_vector(y.mutable_array(), *M);
+
+      // Set BC dofs to zero (effectively zeroes rows of A)
+      fem::set_bc(y.mutable_array(), {bc}, 0.0);
+
+      // Communicate ghost values
+      y.scatter_rev(common::IndexMap::Mode::add);
+    };
+
+    // Compute solution using the conjugate gradient method
+    auto u = std::make_shared<fem::Function<T>>(V);
+    int it = linalg::cg(*u->x(), b, action, 200, 1e-6);
+
+    // Update ghost values and zero y
+    u->x()->scatter_fwd();
+
+    // Set BC values in the solution vectors
+    fem::set_bc(u->x()->mutable_array(), {bc}, 1.0);
+
+    // Compute H1 error e = (u - u_d, u - u_d)*dx
```

Comment:
'L2 error of the solution vector' (which is different from the finite element error).